tests/drivers/adc/adc_api/src/test_adc.c

```diff
@@ -8,98 +9,333 @@
-#define ADC_CHANNEL	1
-#else
-#define ADC_DEV_NAME	CONFIG_ADC_0_NAME
-#define ADC_CHANNEL	10
-#endif
+/* These settings need to be defined separately for particular boards. */
+#define ADC_DEVICE_NAME		CONFIG_ADC_0_NAME
+#define ADC_RESOLUTION		10
+#define ADC_GAIN		ADC_GAIN_1_2
+#define ADC_REFERENCE		ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+#define ADC_1ST_CHANNEL_ID	0
+/*#define ADC_2ND_CHANNEL_ID	2*/
 
-static u16_t seq_buffer[BUFFER_SIZE];
+#define BUFFER_SIZE  6
+static s16_t m_sample_buffer[BUFFER_SIZE];
 
-static struct adc_seq_entry entry = {
-	.sampling_delay = 30,
-	.channel_id = ADC_CHANNEL,
-	.buffer = (void *)seq_buffer,
-	.buffer_length = BUFFER_SIZE * sizeof(seq_buffer[0])
+static const struct adc_channel_cfg m_1st_channel_cfg = {
+	.gain             = ADC_GAIN,
+	.reference        = ADC_REFERENCE,
+	.acquisition_time = ADC_ACQUISITION_TIME,
+	.channel_id       = ADC_1ST_CHANNEL_ID,
+#if defined(CONFIG_ADC_CONFIGURABLE_INPUTS)
+	.input_positive   = ADC_1ST_CHANNEL_INPUT,
+#endif
 };
-
-static struct adc_seq_table table = {
-	.entries = &entry,
-	.num_entries = 1,
+#if defined(ADC_2ND_CHANNEL_ID)
+static const struct adc_channel_cfg m_2nd_channel_cfg = {
+	.gain             = ADC_GAIN,
+	.reference        = ADC_REFERENCE,
+	.acquisition_time = ADC_ACQUISITION_TIME,
+	.channel_id       = ADC_2ND_CHANNEL_ID,
+#if defined(CONFIG_ADC_CONFIGURABLE_INPUTS)
+	.input_positive   = ADC_2ND_CHANNEL_INPUT,
+#endif
 };
+#endif /* defined(ADC_2ND_CHANNEL_ID) */
+
+static struct device *init_adc(void)
+{
+	int ret;
+	struct device *adc_dev = device_get_binding(ADC_DEVICE_NAME);
+
+	zassert_not_null(adc_dev, "Cannot get ADC device");
+
+	ret = adc_channel_setup(adc_dev, &m_1st_channel_cfg);
+	zassert_equal(ret, 0,
+		"Setting up of the first channel failed with code %d", ret);
+
+#if defined(ADC_2ND_CHANNEL_ID)
+	ret = adc_channel_setup(adc_dev, &m_2nd_channel_cfg);
+	zassert_equal(ret, 0,
+		"Setting up of the second channel failed with code %d", ret);
+#endif /* defined(ADC_2ND_CHANNEL_ID) */
+
+	memset(m_sample_buffer, 0, sizeof(m_sample_buffer));
 
-static int test_task(void)
+	return adc_dev;
+}
+
+static void check_samples(int expected_count)
 {
 	int i;
+
+	TC_PRINT("Samples read: ");
+	for (i = 0; i < BUFFER_SIZE; i++) {
+		s16_t sample_value = m_sample_buffer[i];
+
+		TC_PRINT("0x%04x ", sample_value);
+		if (i < expected_count) {
+			zassert_not_equal(0, sample_value,
+				"[%u] should be non-zero", i);
+		} else {
+			zassert_equal(0, sample_value,
+				"[%u] should be zero", i);
+		}
+	}
+	TC_PRINT("\n");
+}
+
+/*******************************************************************************
+ * test_adc_sample_one_channel
+ */
+static int test_task_one_channel(void)
+{
 	int ret;
-	struct device *adc_dev = device_get_binding(ADC_DEV_NAME);
+	const struct adc_sequence sequence = {
+		.channels    = BIT(ADC_1ST_CHANNEL_ID),
+		.buffer      = m_sample_buffer,
+		.buffer_size = sizeof(m_sample_buffer),
```

Comment:
@anangl  `sizeof(m_sample_buffer) ` this returns 4 byte (size of array). Please change it to ` BUFFER_SIZE * sizeof(m_sample_buffer[0])`. We were facing crash while porting designware driver.